File: verifier/reference_model.js
const WAD = 10n ** 18n;
const BPS_DENOMINATOR = 10000n;

function normalizeToWad(amount, decimals) {
  if (decimals === 18) return amount;
  if (decimals < 18) return amount * (10n ** BigInt(18 - decimals));
  return amount / (10n ** BigInt(decimals - 18));
}

function denormalizeFromWad(wadAmount, decimals) {
  if (decimals === 18) return wadAmount;
  if (decimals < 18) return wadAmount / (10n ** BigInt(18 - decimals));
  return wadAmount * (10n ** BigInt(decimals - 18));
}

function evaluateReferenceModel(input) {
  const rawTokenAmount = BigInt(input.rawBalance || input.rawTokenAmount || 0);
  const currentMultiplier = BigInt(input.currentMultiplier || input.effectiveMultiplier || WAD);
  const pendingMultiplier = input.pendingMultiplier ? BigInt(input.pendingMultiplier) : 0n;
  const effectiveAt = input.effectiveAt || 0;
  const blockTimestamp = input.blockTimestamp || Math.floor(Date.now() / 1000);
  const totalReturnPrice8 = BigInt(input.oraclePrice || input.totalReturnPrice8 || 0);
  const debt = BigInt(input.debtAmountUsd || input.totalDebtUsd || 0);
  const maxOracleDelay = BigInt(input.maxOracleDelay || 86400);
  const transitionGuardWindow = BigInt(input.transitionGuardWindow || 3600);
  const guardReductionBps = BigInt(input.guardLtvReductionBps || 500);
  const assetDecimals = input.assetDecimals || 18;
  const oracleDecimals = input.oracleDecimals || 8;

  // 1. Determine live effective UI multiplier (for UI presentation / diagnostics ONLY)
  let effectiveMultiplier = currentMultiplier;
  let hasLivePending = false;
  if (effectiveAt !== 0 && pendingMultiplier !== 0n) {
    if (blockTimestamp >= effectiveAt) {
      effectiveMultiplier = pendingMultiplier;
    } else {
      hasLivePending = true;
    }
  }

  // 2. Compute UI Share-Equivalent Amount
  const uiShareAmount = (rawTokenAmount * effectiveMultiplier) / WAD;

  // 3. Compute Canonical Collateral Valuation under Total Return Value (TRV) Rule:
  // CANONICAL RULE: rawTokenAmount * totalReturnPrice8
  // NEVER compounds effectiveMultiplier into TRV!
  const normRaw = normalizeToWad(rawTokenAmount, assetDecimals);
  const normPrice = normalizeToWad(totalReturnPrice8, oracleDecimals);
  const canonicalCollateralUsd = (normRaw * normPrice) / WAD;

  // 4. Compute Naive Double-Adjusted Valuation (raw * multiplier * TRV)
  const naiveDoubleAdjustedUsd = (canonicalCollateralUsd * effectiveMultiplier) / WAD;
  const doubleAdjustmentDetected = (effectiveMultiplier !== WAD);

  // 5. Check Transition Guard Window
  let activeLtvBps = BigInt(input.ltvBps || 7500);
  let inGuardWindow = false;
  if (hasLivePending && effectiveAt > blockTimestamp) {
    const timeToEffective = BigInt(effectiveAt - blockTimestamp);
    if (timeToEffective <= transitionGuardWindow) {
      inGuardWindow = true;
      if (activeLtvBps > guardReductionBps) {
        activeLtvBps -= guardReductionBps;
      }
    }
  }

  // 6. Max Debt
  const maxDebtUsd = (canonicalCollateralUsd * activeLtvBps) / BPS_DENOMINATOR;

  // 7. Health Factor
  let healthFactor = WAD * 1000n;
  if (debt === 0n) {
    healthFactor = 2n ** 256n - 1n; // max uint256
  } else {
    const liqCollateral = (canonicalCollateralUsd * BigInt(input.liquidationThresholdBps || 8500)) / BPS_DENOMINATOR;
    healthFactor = (liqCollateral * WAD) / debt;
  }

  // 8. Freshness and Status
  const isStale = (BigInt(blockTimestamp) - BigInt(input.oracleUpdatedAt || blockTimestamp)) > maxOracleDelay;

  let status = "COHERENT";
  if (input.isTransferPaused) {
    status = "BLOCKED";
  } else if (isStale) {
    status = "STALE_ORACLE";
  } else if (healthFactor < WAD && debt > 0n) {
    status = "LIQUIDATABLE";
  } else if (inGuardWindow) {
    status = "TRANSITION";
  } else if (hasLivePending) {
    status = "PENDING_ACTION";
  } else {
    status = "COHERENT";
  }

  return {
    valuationBasis: "RAW_X_TOTAL_RETURN",
    multiplierAppliedToValue: false,
    rawTokenAmount,
    effectiveMultiplier,
    uiShareAmount,
    totalReturnPrice8,
    canonicalCollateralUSD: canonicalCollateralUsd,
    naiveDoubleAdjustedUSD: naiveDoubleAdjustedUsd,
    doubleAdjustmentDetected,
    maxDebtUsd,
    totalDebtUsd: debt,
    healthFactor,
    status,
    isHealthy: healthFactor >= WAD,
    isLiquidatable: healthFactor < WAD && debt > 0n,
  };
}

module.exports = {
  normalizeToWad,
  denormalizeFromWad,
  evaluateReferenceModel,
};

File: verifier/reference_model.py
\"\"\"
EQUIVANCE — Clean-Room Reference Model (Python 3)
Deterministic reference model enforcing Valuation-Basis Integrity
\"\"\"

import sys
import json
from typing import Dict, Any

WAD = 10**18
BPS_DENOMINATOR = 10000
MAX_UINT256 = 2**256 - 1

def normalize_to_wad(amount: int, decimals: int) -> int:
    if decimals == 18:
        return amount
    elif decimals < 18:
        return amount * (10 ** (18 - decimals))
    else:
        return amount // (10 ** (decimals - 18))

def evaluate_position(data: Dict[str, Any]) -> Dict[str, Any]:
    raw_balance = int(data.get("rawBalance", data.get("rawTokenAmount", 0)))
    current_multiplier = int(data.get("currentMultiplier", data.get("effectiveMultiplier", WAD)))
    pending_multiplier = int(data.get("pendingMultiplier", 0))
    effective_at = int(data.get("effectiveAt", 0))
    block_timestamp = int(data.get("blockTimestamp", 1750000000))
    oracle_price = int(data.get("oraclePrice", data.get("totalReturnPrice8", 0)))
    oracle_decimals = int(data.get("oracleDecimals", 8))
    asset_decimals = int(data.get("assetDecimals", 18))
    oracle_updated_at = int(data.get("oracleUpdatedAt", block_timestamp))
    max_oracle_delay = int(data.get("maxOracleDelay", 86400))
    ltv_bps = int(data.get("ltvBps", 7500))
    liq_threshold_bps = int(data.get("liquidationThresholdBps", 8500))
    debt_amount = int(data.get("debtAmountUsd", data.get("totalDebtUsd", 0)))
    transition_guard_window = int(data.get("transitionGuardWindow", 3600))
    guard_ltv_reduction_bps = int(data.get("guardLtvReductionBps", 500))
    is_paused = bool(data.get("isTransferPaused", False))

    # 1. Effective multiplier
    effective_multiplier = current_multiplier
    has_live_pending = False
    if effective_at != 0 and pending_multiplier != 0:
        if block_timestamp >= effective_at:
            effective_multiplier = pending_multiplier
        else:
            has_live_pending = True

    # 2. UI Share amount
    ui_share_amount = (raw_balance * effective_multiplier) // WAD

    # 3. Canonical Collateral value under TRV rule (rawBalance * totalReturnPrice)
    norm_raw = normalize_to_wad(raw_balance, asset_decimals)
    norm_price = normalize_to_wad(oracle_price, oracle_decimals)
    canonical_collateral_usd = (norm_raw * norm_price) // WAD

    # 4. Naive double adjusted value
    naive_double_adjusted_usd = (canonical_collateral_usd * effective_multiplier) // WAD
    double_adjustment_detected = (effective_multiplier != WAD)

    # 5. Guard window check
    active_ltv = ltv_bps
    in_guard_window = False
    if has_live_pending and effective_at > block_timestamp:
        if (effective_at - block_timestamp) <= transition_guard_window:
            in_guard_window = True
            if active_ltv > guard_ltv_reduction_bps:
                active_ltv -= guard_ltv_reduction_bps

    # 6. Max debt
    max_debt_usd = (canonical_collateral_usd * active_ltv) // BPS_DENOMINATOR

    # 7. Health factor
    if debt_amount == 0:
        health_factor = MAX_UINT256
    else:
        liq_collateral = (canonical_collateral_usd * liq_threshold_bps) // BPS_DENOMINATOR
        health_factor = (liq_collateral * WAD) // debt_amount

    # 8. Status
    is_stale = (block_timestamp - oracle_updated_at) > max_oracle_delay

    if is_paused:
        status = "BLOCKED"
    elif is_stale:
        status = "STALE_ORACLE"
    elif health_factor < WAD and debt_amount > 0:
        status = "LIQUIDATABLE"
    elif in_guard_window:
        status = "TRANSITION"
    elif has_live_pending:
        status = "PENDING_ACTION"
    else:
        status = "COHERENT"

    return {
        "valuationBasis": "RAW_X_TOTAL_RETURN",
        "multiplierAppliedToValue": False,
        "uiMultiplier": str(effective_multiplier),
        "uiShareAmount": str(ui_share_amount),
        "canonicalCollateralUSD": str(canonical_collateral_usd),
        "naiveDoubleAdjustedUSD": str(naive_double_adjusted_usd),
        "doubleAdjustmentDetected": double_adjustment_detected,
        "maxDebtUSD": str(max_debt_usd),
        "totalDebtUSD": str(debt_amount),
        "healthFactor": str(health_factor),
        "status": "PASS" if status in ["COHERENT", "TRANSITION", "PENDING_ACTION"] else status,
    }

if __name__ == "__main__":
    if len(sys.argv) > 1:
        with open(sys.argv[1], "r") as f:
            payload = json.load(f)
        result = evaluate_position(payload)
        print(json.dumps(result, indent=2))
    else:
        sample = {
            "blockTimestamp": 1750000000,
            "rawBalance": "10000000000000000000",
            "currentMultiplier": "10000000000000000000",
            "pendingMultiplier": "10000000000000000000",
            "effectiveAt": 1750000000,
            "oraclePrice": "20000000000",
            "oracleDecimals": 8,
            "assetDecimals": 18,
            "oracleUpdatedAt": 1750000000,
            "ltvBps": 7500,
            "liquidationThresholdBps": 8500,
            "debtAmountUsd": "1500000000000000000000"
        }
        res = evaluate_position(sample)
        print(json.dumps(res, indent=2))
